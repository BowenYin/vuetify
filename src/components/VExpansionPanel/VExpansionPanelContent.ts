import { VExpandTransition } from '../transitions'

import Bootable from '../../mixins/bootable'
import Toggleable from '../../mixins/toggleable'
import Rippleable from '../../mixins/rippleable'
import { Registrable, inject as RegistrableInject } from '../../mixins/registrable'

import VIcon from '../VIcon'
import {
  VListItem,
  VListItemContent,
  VListItemIcon
} from '../VList'
import VExpansionPanel from './VExpansionPanel'
import { VSubtitle1 } from '../typography'

import mixins, { ExtractVue } from '../../util/mixins'
import Vue, { VNode } from 'vue'

type VExpansionPanelInstance = InstanceType<typeof VExpansionPanel>

interface options extends Vue {
  expansionPanel: VExpansionPanelInstance
}

export default mixins<options &
/* eslint-disable indent */
  ExtractVue<[
    typeof Bootable,
    typeof Toggleable,
    typeof Rippleable,
    Registrable<'expansionPanel'>
  ]>
/* eslint-enable indent */
>(
  Bootable,
  Toggleable,
  Rippleable,
  RegistrableInject('expansionPanel', 'v-expansion-panel-content', 'v-expansion-panel')
  /* @vue/component */
).extend({
  name: 'v-expansion-panel-content',

  props: {
    disabled: Boolean,
    readonly: Boolean,
    expandIcon: {
      type: String,
      default: '$vuetify.icons.expand'
    },
    hideActions: Boolean,
    ripple: {
      type: [Boolean, Object],
      default: false
    },
    title: String
  },

  data: () => ({
    height: 'auto'
  }),

  computed: {
    containerClasses (): object {
      return {
        'v-expansion-panel__container--active': this.isActive,
        'v-expansion-panel__container--disabled': this.isDisabled
      }
    },
    isDisabled (): boolean {
      return this.expansionPanel.disabled || this.disabled
    },
    isReadonly (): boolean {
      return this.expansionPanel.readonly || this.readonly
    }
  },

  beforeMount () {
    this.expansionPanel.register(this)
  },

  beforeDestroy () {
    this.expansionPanel.unregister(this)
  },

  methods: {
    onKeydown (e: KeyboardEvent) {
      // Ensure element is the activeElement
      if (
        e.keyCode === 13 &&
        this.$el === document.activeElement
      ) this.expansionPanel.panelClick(this._uid)
    },
    onClick () {
      this.isReadonly || this.expansionPanel.panelClick(this._uid)
    },
    genBody () {
      return this.$createElement('div', {
        ref: 'body',
        class: 'v-expansion-panel__body',
        directives: [{
          name: 'show',
          value: this.isActive
        }] as any
      }, this.showLazyContent(this.$slots.default))
    },
    genActivator () {
      const children = [this.genListItem()]

      if (
        !this.hideActions &&
        !this.$slots.activator
      ) children.push(this.genIcon())

      return this.$createElement('div', {
        staticClass: 'v-expansion-panel__activator',
        directives: [{
          name: 'ripple',
          value: this.ripple
        }] as any,
        on: { click: this.onClick }
      }, children)
    },
    genIcon () {
      const icon = this.$slots.actions ||
        [this.$createElement(VIcon, this.expandIcon)]

      return this.$createElement('transition', {
        attrs: { name: 'fade-transition' }
      }, [
        this.$createElement(VListItemIcon, {
          staticClass: 'v-expansion-panel__icon',
          directives: [{
            name: 'show',
            value: !this.isDisabled
          }] as any
        }, icon)
      ])
    },
    genListItem () {
      if (this.$slots.activator) return this.$slots.activator

      const h = this.$createElement

      return h(VListItem, [
        this.$slots.header ||
        (this.$scopedSlots.header && this.$scopedSlots.header({ open: this.isActive })) ||
        h(VListItemContent, [h(VSubtitle1, this.title)])
      ])
    },
    toggle (active: boolean) {
      if (active) this.isBooted = true

      this.isActive = active
    }
  },

  render (h): VNode {
    return h('div', {
      staticClass: 'v-expansion-panel__container',
      class: this.containerClasses,
      attrs: {
        tabindex: this.isReadonly || this.isDisabled ? null : 0
      },
      on: { keydown: this.onKeydown }
    }, [
      this.genActivator(),
      h(VExpandTransition, [this.genBody()])
    ])
  }
})
