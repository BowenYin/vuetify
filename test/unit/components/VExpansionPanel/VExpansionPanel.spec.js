import Vue from 'vue'
import { test } from '@/test'
import VExpansionPanel from '@/components/VExpansionPanel/VExpansionPanel'
import VExpansionPanelContent from '@/components/VExpansionPanel/VExpansionPanelContent'

const createPanel = props => {
  return Vue.component('test', {
    render: h => {
      const panelContent = h(VExpansionPanelContent, [
        h('div', { slot: 'header' }, 'header'),
        'content'
      ])
      return h(VExpansionPanel, { props }, [panelContent])
    }
  })
}

test('VExpansionPanel.js', ({ mount, compileToFunctions }) => {
  it('should render component and match snapshot', async () => {
    const wrapper = mount(createPanel())

    expect(wrapper.html()).toMatchSnapshot()

    wrapper.find('.v-expansion-panel__activator')[0].trigger('click')
    await wrapper.vm.$nextTick()
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render a multiple component and match snapshot', async () => {
    const wrapper = mount(createPanel({
      multiple: true
    }))

    expect(wrapper.html()).toMatchSnapshot()

    wrapper.find('.v-expansion-panel__activator')[0].trigger('click')
    await wrapper.vm.$nextTick()
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should show content on v-model change', async () => {
    const wrapper = mount(VExpansionPanel, {
      slots: {
        default: VExpansionPanelContent.options
      }
    })

    expect(wrapper.html()).toMatchSnapshot()

    wrapper.setProps({ value: 0 })
    await wrapper.vm.$nextTick()

    expect(wrapper.first('.v-expansion-panel__container--active')).not.toBe(null)
  })

  it('should show content on mount using v-model', async () => {
    const wrapper = mount(VExpansionPanel, {
      propsData: {
        value: 0
      },
      slots: {
        default: VExpansionPanelContent.options
      }
    })

    await wrapper.vm.$nextTick()
    expect(wrapper.first('.v-expansion-panel__container--active')).not.toBe(null)
  })

  it('should allow array v-model when using multiple prop', async () => {
    const wrapper = mount(VExpansionPanel, {
      propsData: {
        multiple: true,
        value: [1, 3]
      },
      slots: {
        default: [VExpansionPanelContent, VExpansionPanelContent, VExpansionPanelContent, VExpansionPanelContent]
      }
    })

    await wrapper.vm.$nextTick()
    expect(wrapper.find('.v-expansion-panel__container--active').length).toBe(2)
  })

  it('should reset v-model when disabling multiple', async () => {
    const fn = jest.fn()
    const wrapper = mount(VExpansionPanel, {
      propsData: {
        multiple: true,
        value: [true, true]
      },
      slots: {
        default: [VExpansionPanelContent, VExpansionPanelContent]
      }
    })

    wrapper.instance().$on('input', fn)

    await wrapper.vm.$nextTick()
    expect(wrapper.find('.v-expansion-panel__container--active').length).toBe(2)

    wrapper.setProps({ multiple: false })
    await wrapper.vm.$nextTick()
    expect(fn).toHaveBeenCalledWith(null)
  })

  it('should reset v-model when disabling multiple', async () => {
    const input = jest.fn()
    const wrapper = mount(VExpansionPanel, {
      propsData: {
        multiple: true,
        value: [false]
      },
      slots: {
        default: [VExpansionPanelContent]
      }
    })

    wrapper.instance().$on('input', input)

    await wrapper.vm.$nextTick()
    expect(wrapper.find('.v-expansion-panel__container--active').length).toBe(0)

    wrapper.setProps({ multiple: false })
    await wrapper.vm.$nextTick()
    expect(input).toHaveBeenCalledWith(null)
  })

  it('should keep a single item open when disabling multiple', async () => {
    const input = jest.fn()
    const wrapper = mount(VExpansionPanel, {
      propsData: {
        multiple: true,
        value: [false, true]
      },
      slots: {
        default: [VExpansionPanelContent, VExpansionPanelContent]
      }
    })
    wrapper.vm.$on('input', input)

    wrapper.setProps({ multiple: false })
    await wrapper.vm.$nextTick()
    expect(input).toHaveBeenCalledWith(1)
  })

  it('should keep the current item open when enabling multiple', async () => {
    const input = jest.fn()
    const wrapper = mount(VExpansionPanel, {
      propsData: {
        multiple: false,
        value: 1
      },
      slots: {
        default: [VExpansionPanelContent, VExpansionPanelContent]
      }
    })
    wrapper.vm.$on('input', input)

    wrapper.setProps({ multiple: true })
    await wrapper.vm.$nextTick()
    expect(input).toHaveBeenCalledWith([false, true])
    expect(wrapper.find('.v-expansion-panel__container--active').length).toBe(1)
  })

  it('should unregister content', async () => {
    const wrapper = mount(VExpansionPanel, {
      slots: {
        default: [VExpansionPanelContent]
      }
    })

    const content = wrapper.first(VExpansionPanelContent.options)

    expect(wrapper.vm.items.length).toBe(1)

    content.destroy()

    expect(wrapper.vm.items.length).toBe(0)
  })

  it('should update from value', async () => {
    const updatePanels = jest.fn()
    const wrapper = mount(VExpansionPanel, {
      methods: { updatePanels }
    })

    wrapper.vm.updateFromValue([true])

    expect(updatePanels).not.toBeCalled()

    wrapper.vm.updateFromValue(0)

    expect(updatePanels).toBeCalledWith([true])

    wrapper.vm.updateFromValue('foo')

    expect(updatePanels).toBeCalledWith('foo')

    wrapper.vm.updateFromValue(null)

    expect(updatePanels).toBeCalledWith([])
  })
})
