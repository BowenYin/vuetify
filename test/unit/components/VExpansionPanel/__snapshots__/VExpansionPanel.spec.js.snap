// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VExpansionPanel.js should render a multiple component and match snapshot 1`] = `

<div class="v-expansion-panel theme--light">
  <div tabindex="0"
       class="v-expansion-panel__container"
  >
    <div class="v-expansion-panel__activator">
      <div class="v-list__item">
        <div>
          header
        </div>
      </div>
      <div class="v-list__item__icon v-expansion-panel__icon">
        <i aria-hidden="true"
           class="v-icon material-icons theme--light"
        >
          keyboard_arrow_down
        </i>
      </div>
    </div>
    <div class="v-expansion-panel__body"
         style="display: none;"
    >
      content
    </div>
  </div>
</div>

`;

exports[`VExpansionPanel.js should render a multiple component and match snapshot 2`] = `

<div class="v-expansion-panel theme--light">
  <div tabindex="0"
       class="v-expansion-panel__container v-expansion-panel__container--active"
  >
    <div class="v-expansion-panel__activator">
      <div class="v-list__item">
        <div>
          header
        </div>
      </div>
      <div class="v-list__item__icon v-expansion-panel__icon">
        <i aria-hidden="true"
           class="v-icon material-icons theme--light"
        >
          keyboard_arrow_down
        </i>
      </div>
    </div>
    <div class="v-expansion-panel__body expand-transition-enter expand-transition-enter-active"
         style="overflow: hidden; height: 0px;"
    >
      content
    </div>
  </div>
</div>

`;

exports[`VExpansionPanel.js should render component and match snapshot 1`] = `

<div class="v-expansion-panel theme--light">
  <div tabindex="0"
       class="v-expansion-panel__container"
  >
    <div class="v-expansion-panel__activator">
      <div class="v-list__item">
        <div>
          header
        </div>
      </div>
      <div class="v-list__item__icon v-expansion-panel__icon">
        <i aria-hidden="true"
           class="v-icon material-icons theme--light"
        >
          keyboard_arrow_down
        </i>
      </div>
    </div>
    <div class="v-expansion-panel__body"
         style="display: none;"
    >
      content
    </div>
  </div>
</div>

`;

exports[`VExpansionPanel.js should render component and match snapshot 2`] = `

<div class="v-expansion-panel theme--light">
  <div tabindex="0"
       class="v-expansion-panel__container v-expansion-panel__container--active"
  >
    <div class="v-expansion-panel__activator">
      <div class="v-list__item">
        <div>
          header
        </div>
      </div>
      <div class="v-list__item__icon v-expansion-panel__icon">
        <i aria-hidden="true"
           class="v-icon material-icons theme--light"
        >
          keyboard_arrow_down
        </i>
      </div>
    </div>
    <div class="v-expansion-panel__body expand-transition-enter expand-transition-enter-active"
         style="overflow: hidden; height: 0px;"
    >
      content
    </div>
  </div>
</div>

`;

exports[`VExpansionPanel.js should show content on v-model change 1`] = `

<div class="v-expansion-panel theme--light">
  <div tabindex="0"
       class="v-expansion-panel__container"
  >
    <div class="v-expansion-panel__activator">
      <div class="v-list__item">
        <div class="v-list__item__content">
          <div class="subtitle-1">
          </div>
        </div>
      </div>
      <div class="v-list__item__icon v-expansion-panel__icon">
        <i aria-hidden="true"
           class="v-icon material-icons theme--light"
        >
          keyboard_arrow_down
        </i>
      </div>
    </div>
    <div class="v-expansion-panel__body"
         style="display: none;"
    >
    </div>
  </div>
</div>

`;
